File: WeightTracker.java
// WeightTracker.java
import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.util.*;
import com.google.gson.*;

class Config {
    double height = 0.0;
}

class Entry {
    String date;
    double weight;
    String note;
}

class WeightTracker {
    private static final String CONFIG_FILE = "weight_config.json";
    private static final String DATA_FILE = "weights.json";
    private static double height = 0.0;
    private static List<Entry> entries = new ArrayList<>();
    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    static void loadConfig() {
        try {
            Path path = Paths.get(CONFIG_FILE);
            if (Files.exists(path)) {
                String json = new String(Files.readAllBytes(path));
                Config cfg = gson.fromJson(json, Config.class);
                height = cfg.height;
            }
        } catch (Exception e) {}
    }

    static void saveConfig() {
        try {
            Config cfg = new Config();
            cfg.height = height;
            Files.write(Paths.get(CONFIG_FILE), gson.toJson(cfg).getBytes());
        } catch (Exception e) {}
    }

    static void loadEntries() {
        try {
            Path path = Paths.get(DATA_FILE);
            if (Files.exists(path)) {
                String json = new String(Files.readAllBytes(path));
                Entry[] arr = gson.fromJson(json, Entry[].class);
                entries = new ArrayList<>(Arrays.asList(arr));
            }
        } catch (Exception e) {}
    }

    static void saveEntries() {
        try {
            Files.write(Paths.get(DATA_FILE), gson.toJson(entries).getBytes());
        } catch (Exception e) {}
    }

    static Double bmi(double weight) {
        if (height <= 0) return null;
        double h = height / 100.0;
        return weight / (h * h);
    }

    static String bmiCategory(double bmiVal) {
        if (bmiVal < 18.5) return "Underweight";
        if (bmiVal < 25.0) return "Normal";
        if (bmiVal < 30.0) return "Overweight";
        return "Obese";
    }

    static void setHeight(double h) {
        height = h;
        saveConfig();
        System.out.printf("✅ Height set to %.1f cm\n", h);
    }

    static void addEntry(double weight, String date, String note) {
        if (date == null) date = LocalDate.now().toString();
        if (note == null) note = "";
        Entry e = new Entry();
        e.date = date;
        e.weight = weight;
        e.note = note;
        entries.add(e);
        saveEntries();
        Double bmiVal = bmi(weight);
        String bmiStr = bmiVal != null ? String.format(" (BMI: %.1f)", bmiVal) : "";
        System.out.printf("✅ Logged: %.1f kg on %s%s\n", weight, date, bmiStr);
    }

    static void listEntries() {
        if (entries.isEmpty()) {
            System.out.println("No entries.");
            return;
        }
        System.out.println("\n📋 Weight Log:");
        for (Entry e : entries) {
            Double bmiVal = bmi(e.weight);
            String bmiStr = bmiVal != null ? String.format("BMI: %.1f", bmiVal) : "N/A";
            String note = e.note != null && !e.note.isEmpty() ? " | " + e.note : "";
            System.out.printf("%s | %.1f kg | %s%s\n", e.date, e.weight, bmiStr, note);
        }
    }

    static void stats() {
        if (entries.isEmpty()) {
            System.out.println("No entries.");
            return;
        }
        if (height <= 0) {
            System.out.println("❌ Please set your height first: height <cm>");
            return;
        }
        int total = entries.size();
        double sum = 0, min = Double.MAX_VALUE, max = Double.MIN_VALUE;
        for (Entry e : entries) {
            sum += e.weight;
            if (e.weight < min) min = e.weight;
            if (e.weight > max) max = e.weight;
        }
        double avg = sum / total;
        double current = entries.get(total - 1).weight;
        Double bmiVal = bmi(current);
        String cat = bmiVal != null ? bmiCategory(bmiVal) : "N/A";
        System.out.printf("\n⚖️ Weight Tracker\n");
        System.out.printf("Height: %.1f cm\n", height);
        System.out.printf("\n📊 Statistics\n");
        System.out.printf("Entries: %d\n", total);
        System.out.printf("Current: %.1f kg\n", current);
        System.out.printf("Average: %.1f kg\n", avg);
        System.out.printf("Min: %.1f kg\n", min);
        System.out.printf("Max: %.1f kg\n", max);
        if (bmiVal != null) {
            System.out.printf("\n📐 BMI: %.1f (%s)\n", bmiVal, cat);
            System.out.println("BMI Range: 18.5 - 24.9");
        }
    }

    static void exportCSV(String filename) throws IOException {
        Path path = Paths.get(filename);
        try (BufferedWriter writer = Files.newBufferedWriter(path)) {
            writer.write("Date,Weight (kg),BMI,Note\n");
            for (Entry e : entries) {
                Double bmiVal = bmi(e.weight);
                String bmiStr = bmiVal != null ? String.format("%.1f", bmiVal) : "";
                writer.write(String.format("%s,%.1f,%s,%s\n", e.date, e.weight, bmiStr, e.note != null ? e.note : ""));
            }
        }
        System.out.printf("✅ Exported %d entries to %s\n", entries.size(), filename);
    }

    public static void main(String[] args) throws Exception {
        loadConfig();
        loadEntries();
        if (args.length < 1) {
            System.out.println("Usage: WeightTracker [height|add|list|stats|export]");
            return;
        }
        String cmd = args[0];
        switch (cmd) {
            case "height":
                if (args.length < 2) { System.out.println("Usage: height <cm>"); return; }
                setHeight(Double.parseDouble(args[1]));
                break;
            case "add":
                if (args.length < 2) { System.out.println("Usage: add <weight> [--date YYYY-MM-DD] [--note TEXT]"); return; }
                double weight = Double.parseDouble(args[1]);
                String date = null, note = null;
                for (int i = 2; i < args.length; i++) {
                    if (args[i].equals("--date") && i+1 < args.length) date = args[++i];
                    if (args[i].equals("--note") && i+1 < args.length) note = args[++i];
                }
                addEntry(weight, date, note);
                break;
            case "list":
                listEntries();
                break;
            case "stats":
                stats();
                break;
            case "export":
                String filename = args.length > 1 ? args[1] : "weights.csv";
                exportCSV(filename);
                break;
            default:
                System.out.println("Unknown command");
        }
    }
}

File: weight_tracker.go
// weight_tracker.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"time"
)

type Config struct {
	Height float64 `json:"height"`
}

type Entry struct {
	Date   string  `json:"date"`
	Weight float64 `json:"weight"`
	Note   string  `json:"note,omitempty"`
}

type Tracker struct {
	Height  float64   `json:"height"`
	Entries []Entry   `json:"entries"`
	Config  *Config   `json:"-"`
}

var configFile = "weight_config.json"
var dataFile = "weights.json"

func loadConfig() Config {
	var cfg Config
	data, err := os.ReadFile(configFile)
	if err != nil {
		return cfg
	}
	json.Unmarshal(data, &cfg)
	return cfg
}

func saveConfig(cfg Config) {
	data, _ := json.MarshalIndent(cfg, "", "  ")
	os.WriteFile(configFile, data, 0644)
}

func loadEntries() []Entry {
	var entries []Entry
	data, err := os.ReadFile(dataFile)
	if err != nil {
		return entries
	}
	json.Unmarshal(data, &entries)
	return entries
}

func saveEntries(entries []Entry) {
	data, _ := json.MarshalIndent(entries, "", "  ")
	os.WriteFile(dataFile, data, 0644)
}

func bmi(weight, height float64) float64 {
	if height <= 0 {
		return 0
	}
	h := height / 100.0
	return weight / (h * h)
}

func bmiCategory(bmi float64) string {
	if bmi < 18.5 {
		return "Underweight"
	} else if bmi < 25.0 {
		return "Normal"
	} else if bmi < 30.0 {
		return "Overweight"
	} else {
		return "Obese"
	}
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: weight_tracker [height|add|list|stats|export]")
		return
	}
	cfg := loadConfig()
	entries := loadEntries()
	cmd := os.Args[1]

	switch cmd {
	case "height":
		if len(os.Args) != 3 {
			fmt.Println("Usage: height <cm>")
			return
		}
		var h float64
		fmt.Sscanf(os.Args[2], "%f", &h)
		cfg.Height = h
		saveConfig(cfg)
		fmt.Printf("✅ Height set to %.1f cm\n", h)

	case "add":
		if len(os.Args) < 3 {
			fmt.Println("Usage: add <weight> [--date YYYY-MM-DD] [--note TEXT]")
			return
		}
		var w float64
		fmt.Sscanf(os.Args[2], "%f", &w)
		date := time.Now().Format("2006-01-02")
		note := ""
		for i := 3; i < len(os.Args); i++ {
			if os.Args[i] == "--date" && i+1 < len(os.Args) {
				date = os.Args[i+1]
				i++
			}
			if os.Args[i] == "--note" && i+1 < len(os.Args) {
				note = os.Args[i+1]
				i++
			}
		}
		entry := Entry{Date: date, Weight: w, Note: note}
		entries = append(entries, entry)
		saveEntries(entries)
		bmiVal := bmi(w, cfg.Height)
		if bmiVal > 0 {
			fmt.Printf("✅ Logged: %.1f kg on %s (BMI: %.1f)\n", w, date, bmiVal)
		} else {
			fmt.Printf("✅ Logged: %.1f kg on %s\n", w, date)
		}

	case "list":
		if len(entries) == 0 {
			fmt.Println("No entries.")
			return
		}
		fmt.Println("\n📋 Weight Log:")
		for _, e := range entries {
			bmiVal := bmi(e.Weight, cfg.Height)
			bmiStr := "N/A"
			if bmiVal > 0 {
				bmiStr = fmt.Sprintf("%.1f", bmiVal)
			}
			note := e.Note
			if note != "" {
				note = " | " + note
			}
			fmt.Printf("%s | %.1f kg | BMI: %s%s\n", e.Date, e.Weight, bmiStr, note)
		}

	case "stats":
		if len(entries) == 0 {
			fmt.Println("No entries.")
			return
		}
		if cfg.Height <= 0 {
			fmt.Println("❌ Please set your height first: height <cm>")
			return
		}
		total := len(entries)
		sum := 0.0
		min := entries[0].Weight
		max := entries[0].Weight
		for _, e := range entries {
			sum += e.Weight
			if e.Weight < min {
				min = e.Weight
			}
			if e.Weight > max {
				max = e.Weight
			}
		}
		avg := sum / float64(total)
		current := entries[total-1].Weight
		bmiVal := bmi(current, cfg.Height)
		cat := bmiCategory(bmiVal)
		fmt.Printf("\n⚖️ Weight Tracker\n")
		fmt.Printf("Height: %.1f cm\n", cfg.Height)
		fmt.Printf("\n📊 Statistics\n")
		fmt.Printf("Entries: %d\n", total)
		fmt.Printf("Current: %.1f kg\n", current)
		fmt.Printf("Average: %.1f kg\n", avg)
		fmt.Printf("Min: %.1f kg\n", min)
		fmt.Printf("Max: %.1f kg\n", max)
		if bmiVal > 0 {
			fmt.Printf("\n📐 BMI: %.1f (%s)\n", bmiVal, cat)
			fmt.Println("BMI Range: 18.5 - 24.9")
		}

	case "export":
		filename := "weights.csv"
		if len(os.Args) >= 3 {
			filename = os.Args[2]
		}
		f, err := os.Create(filename)
		if err != nil {
			fmt.Println("Error creating file:", err)
			return
		}
		defer f.Close()
		f.WriteString("Date,Weight (kg),BMI,Note\n")
		for _, e := range entries {
			bmiVal := bmi(e.Weight, cfg.Height)
			bmiStr := ""
			if bmiVal > 0 {
				bmiStr = fmt.Sprintf("%.1f", bmiVal)
			}
			f.WriteString(fmt.Sprintf("%s,%.1f,%s,%s\n", e.Date, e.Weight, bmiStr, e.Note))
		}
		fmt.Printf("✅ Exported %d entries to %s\n", len(entries), filename)

	default:
		fmt.Println("Unknown command")
	}
}
